pkg/api/types.go

```diff
@@ -1736,8 +1736,17 @@ type Taint struct {
 	Value string
 	// Required. The effect of the taint on pods
 	// that do not tolerate the taint.
-	// Valid effects are NoSchedule and PreferNoSchedule.
+	// Valid effects are NoSchedule, PreferNoSchedule and NoExecute.
 	Effect TaintEffect
+	// TimeAdded represents the time at which the taint was added.
+	// Taint's effect must be NoExecute, otherwise this field is ignored.
+	// Toleration with tolerationSeconds will tolerate the taint for
+	// only a duration (indicated with toleration.tolerationSeconds)
+	// that starts at timeAdded. By default, taint.timeAdded is not set,
+	// which means the taint can only be tolerated by toleration that
+	// tolerates infinite duration.
+	// +optional
+	TimeAdded metav1.Time
```

Comment:
I'm not sure it'll be useful - when making eviction decisions we won't be able to depend on this timestamp either way (clock skew). OTOH it's probably useful for usability perspective, so we can show user when given taint was added, but then comment is wrong (it has the same sense for NoExecute taint as for other ones). @davidopp